### DISCUSSION TITLE
Pipe console.* output

### DISCUSSION THREAD

**Discussion by martsokha** [rscarson/rustyscript#189] [General]
Is there a way to pipe `console.log` output into `tracing::info`?

**Comment by rscarson** [comment]
I think the easiest and most performant option would be a small extension:

This is a small extension that provides a javascript-binding for `tracing::info`:
```rust
#[op2(fast)]
fn op_trace(#[string] name: &str, #[string] msg: &str) {
    tracing::info!(name: name, msg);
}
extension!(
    impl_trace,
    ops = [op_trace],
    esm_entry_point = "ext:impl_trace/trace.js",
    esm = [ dir "impl_trace", "trace.js" ],
);
```

And `impl_trace/trace.js`, which should stub `console.log` with the required functionality
```javascript
_log = globalThis.console.log;
globalThis.console.log = (msg) => {
    Deno.core.ops.op_trace("console.log", msg);
    _log(msg);
};
```

**Reply by martsokha** [reply]
Thanks for the response! 
I got another question, what's the proper way to convert `v8::Global<v8::Function>` into `js_value::Function`?
I'm trying to implement something similar to `Deno.test` i.e. to register a js function and then call it from rust.

```rust
#[op2]
#[string]
pub fn register_service(
    #[state] registry: &Registry,
    #[serde] options: Options,
    #[global] func: v8::Global<v8::Function>,
) -> Result<String> {
    // ...
    let id = registry.register_function(options, func)?;
    Ok(id.to_string())
}
```

and then

```rust
pub fn execute(&self) -> Result<()> {
    // ...
    let func /* v8::Global<v8::Function> */ = self.registry.find(id)?;
    let _ = self.runtime.call_stored_function_async(???).await?;
    // ...
}
```

**Reply by rscarson** [reply]
That is a bit of a blindspot in the current version I'm afraid

I've just added `try_from_v8` and `from_v8_unchecked` for this use-case

The 2 versions of this you could use are:

The safer, checked version - which requires a HandleScope:
```rust
Function::try_from_v8(&mut runtime.deno_runtime().handle_scope(), global)
```

Or an unchecked variant that is safe if you are checking the value type yourself - which here you are!
```rust
unsafe { Function::from_v8_unchecked(global) }
```

**If you do use the unsafe variant always guard for the type of value, or the runtime could panic when you actually try to call the function!**

**Reply by martsokha** [reply]
Alright, thanks again! I'll try tomorrow.

**Reply by rscarson** [reply]
v0.8.0 is now published with the changes described
